presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionOptimizer.java

```diff
@@ -24,31 +26,38 @@
 
 import static com.facebook.presto.spi.relation.ExpressionOptimizer.Level.OPTIMIZED;
 import static com.facebook.presto.sql.planner.LiteralEncoder.toRowExpression;
+import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 public final class RowExpressionOptimizer
         implements ExpressionOptimizer
 {
-    private final Metadata metadata;
+    private final FunctionAndTypeManager functionAndTypeManager;
 
     public RowExpressionOptimizer(Metadata metadata)
     {
-        this.metadata = requireNonNull(metadata, "metadata is null");
+        this(requireNonNull(metadata, "metadata is null").getFunctionAndTypeManager());
+    }
+
+    public RowExpressionOptimizer(FunctionMetadataManager functionMetadataManager)
+    {
+        checkArgument(functionMetadataManager instanceof FunctionAndTypeManager, "Expected functionMetadataManager to be instance of FunctionAndTypeManager");
+        this.functionAndTypeManager = (FunctionAndTypeManager) requireNonNull(functionMetadataManager, "functionMetadataManager is null");
     }
```

Comment:
This check seems unnecessary,

```suggestion

    public RowExpressionOptimizer(FunctionAndTypeManager functionAndTypeManager)
    {
        this.functionAndTypeManager =  requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
    }
```

Reply:
Good catch, done.